runtests.sh

```diff
@@ -79,50 +81,64 @@ if [ ! -d "$INVALID" ]; then
     exit 1
 fi
 
-# Run the giftcardreader on each of the testcases. Valid gift cards
+# Run the giftcardreader versions on each of the testcases. Valid gift cards
 # should return 0, invalid gift cards should return non-zero. Crashes
 # are detected by checking for a return value greater than or equal to
 # 128, and are never considered a pass.
+# Credit to Clark Cone (NYU Student) for improving the count function
 passed=0
 failed=0
 echo "Running tests on ${BOLD}valid${RESET} gift cards (expected return value: 0)..."
 if exists "$VALID"/* ; then
-    printf "${BOLD}%-50s %-5s %s${RESET}\n" "Testcase" "Pass?" "Exit Status"
+    printf "${BOLD}%-50s %-5s %-20s %s${RESET}\n" "Testcase" "Pass?" "Version" "Exit Status"
     for gft in "$VALID"/* ; do
         [ -f "$gft" ] || break
-        timeout $MAX_RUNTIME ./giftcardreader.original 1 "$gft" &> /dev/null
-        rv=$?
-        if [ $rv -eq 0 ]; then
-            pmsg="PASS"
-            passed=$((passed+1))
-        else
-            pmsg="FAIL"
-            failed=$((failed+1))
-        fi
-        [ "$pmsg" = "PASS" ] && rowcolor=$GREEN || rowcolor=$RED
-        printf "${rowcolor}%-50s %-5s %s${RESET}\n" "$(basename "$gft")" "$pmsg" "$(check_exit_status $rv)"
+        test_passed=false
+        for exe in giftcardreader.original giftcardreader.ubsan giftcardreader.asan; do
+            timeout $MAX_RUNTIME ./$exe 1 "$gft" &> /dev/null
+            rv=$?
+            if [ $rv -eq 0 ]; then
+                pmsg="PASS"
+                test_passed=true
+                rowcolor=$GREEN
+                printf "${rowcolor}%-50s %-5s %-20s %s${RESET}\n" "$(basename "$gft")" "$pmsg" "$exe" "$(check_exit_status $rv)"
+                passed=$((passed+1))
+                break
```

Comment:
The break statement here causes the script to stop testing the remaining binaries (asan and ubsan) once .original passes. This results in the test case being marked as PASS even if asan or ubsan crash. I encountered this issue when a student's gift card crashed asan and ubsan but not .original, as the .original binary is compiled without runtime checks and does not detect out-of-bounds memory access or undefined behavior.